server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -298,34 +379,126 @@ public SubView(final BarrageStreamGenerator generator,
-                        this, offset, offset + batchSize, metadata, generator::appendModColumns);
+                        this, 0, 0, metadata, generator::appendAddColumns);
                 bytesWritten += is.available();
                 visitor.accept(is);
-                offset += batchSize;
-                metadata = null;
+            } else {
+                long offset = 0;
+                long chunkOffset = 0;
+
+                while (offset < numAddRows) {
+                    long rowsRemaining = numAddRows - offset;
+
+                    int batchSize = (int) Math.min(rowsRemaining, batchSizeLimit);
+
+                    // make sure we are not about to cross a chunk boundary
+                    if (chunkOffset + batchSize > TWO_DIMENSIONAL_COLUMN_SOURCE_THRESHOLD) {
+                        batchSize = (int) (TWO_DIMENSIONAL_COLUMN_SOURCE_THRESHOLD - chunkOffset);
+                        chunkOffset = 0;
+                    }
+
+                    try {
+                        final InputStream is = generator.getInputStream(
+                                this, offset, offset + batchSize, metadata, generator::appendAddColumns);
+                        int bytesToWrite = is.available();
+
+                        // treat this as a hard limit, exceeding fails a client or w2w (unless we are sending a single
+                        // row then we must send and let it potentially fail)
+                        if (bytesToWrite > DEFAULT_MESSAGE_SIZE_LIMIT && batchSize > 1) {
+                            // can't write this, so close the input stream and retry
+                            is.close();
+                        } else {
+                            bytesWritten += bytesToWrite;
+
+                            // let's write the data
+                            visitor.accept(is);
+
+                            offset += batchSize;
+                            chunkOffset += batchSize;
+                            metadata = null;
+                        }
+                        // recompute the batch limit for the next message
+                        int bytesPerRow = bytesToWrite / batchSize;
+                        if (bytesPerRow > 0) {
+                            int rowLimit = DEFAULT_MESSAGE_SIZE_LIMIT / bytesPerRow;
+
+                            // add some margin for abnormal cell contents
+                            batchSizeLimit = Math.min(batchSize(), Math.max(1, (int) ((double) rowLimit * 0.9)));
+                        }
+                    } catch (SizeException ex) {
+                        // was an overflow in the ChunkInputStream generator (probably VarBinary). We can't compute the
+                        // correct number of rows from this failure, so cut batch size in half and try again. This may
+                        // occur multiple times until the size is restricted properly
+                        batchSizeLimit = Math.max(1, batchSizeLimit / 2);
```

Comment:
I just realized that this Math.max check leads to an infinite loop. If we can't send a single row (i.e. batchSizeLimit / 2 == 0) because of size constraints, then we are just simply SOL and we'll need to kill affected subscriptions.